tests/mysql_test.go

```diff
@@ -120,3 +128,138 @@ func TestCheckServerVersionAndBinlogForPITR(t *testing.T) {
+	mysqlDriver, ok := driver.(*pluginmysql.Driver)
+	a.Equal(true, ok)
+	utilDir := t.TempDir()
+	utilInstance, err := mysqlutil.Install(utilDir)
+	a.NoError(err)
+	binlogDir := t.TempDir()
+	mysqlRestore := restoremysql.New(mysqlDriver, utilInstance, connCfg, binlogDir)
+
+	// init schema
+	_, err = db.ExecContext(ctx, `
+		CREATE DATABASE test;
+		USE test;
+		CREATE TABLE tbl (id int);
+		`)
+	a.NoError(err)
+
+	// Rotate to create multiple binlog files for test.
+	var startTsList []int64
+	numRotate := 10
+	// There'll be `numBinlogFiles` binlog files generated, and the last one contains no actual data.
+	numBinlogFiles := numRotate + 1
+	for i := 0; i < numRotate; i++ {
+		// Record the start event ts of the current binlog file.
+		startTsList = append(startTsList, time.Now().Unix())
+		// Insert some data to grow the current binlog file.
+		_, err = db.ExecContext(ctx, fmt.Sprintf(`
+			USE test;
+			INSERT INTO tbl VALUES (%d);
+			`, i))
+		a.NoError(err)
+		// Make a one-second gap between the two binlog files' events
+		time.Sleep(time.Second)
+		// Rotate the binlog file.
+		_, err = db.ExecContext(ctx, "FLUSH BINARY LOGS;")
+		a.NoError(err)
+	}
+	t.Logf("startTsList: %v\n", startTsList)
+
+	binlogFilesOnServer, err := mysqlRestore.GetSortedBinlogFilesMetaOnServer(ctx)
+	a.NoError(err)
+
+	t.Log("Download binlog files in empty dir up to targetTs")
+	for i, targetTs := range startTsList {
+		t.Logf("Round %d\n", i)
+		binlogFilesBefore, err := ioutil.ReadDir(binlogDir)
+		a.NoError(err)
+		for _, file := range binlogFilesBefore {
+			path := filepath.Join(binlogDir, file.Name())
+			err = os.Remove(path)
+			a.NoError(err)
+		}
+		err = mysqlRestore.FetchBinlogFilesUpToTargetTs(ctx, targetTs)
+		a.NoError(err)
+		binlogFilesDownloaded, err := ioutil.ReadDir(binlogDir)
+		a.NoError(err)
+		// We will always download one more file to find out that it exceeds the targetTs.
+		num := (i + 1) + 1
+		if num > numBinlogFiles {
+			num = numBinlogFiles
+		}
+		a.Equal(num, len(binlogFilesDownloaded))
+		for j := range binlogFilesDownloaded {
+			a.Equal(binlogFilesOnServer[j].Name, binlogFilesDownloaded[j].Name())
+			a.Equal(binlogFilesOnServer[j].Size, binlogFilesDownloaded[j].Size())
+		}
+	}
+
+	t.Log("Clean up binlog dir")
+	binlogFiles, err := ioutil.ReadDir(binlogDir)
+	a.NoError(err)
+	for _, file := range binlogFiles {
+		path := filepath.Join(binlogDir, file.Name())
+		err = os.Remove(path)
+		a.NoError(err)
+	}
+
+	t.Log("Truncate or delete downloaded files and re-download")
+	rand.Seed(time.Now().Unix())
+	for i, targetTs := range startTsList {
+		t.Logf("Round %d\n", i)
+		// Fetch and randomly truncate/delete some binlog files.
+		t.Log("Fetch binlog files to targetTs")
+		err = mysqlRestore.FetchBinlogFilesUpToTargetTs(ctx, targetTs)
+		a.NoError(err)
+		binlogFilesDownloaded, err := ioutil.ReadDir(binlogDir)
+		t.Logf("Downloaded %d files to empty dir", len(binlogFilesDownloaded))
```

Comment:
It looks like the directory is empty only on the first round. I guess that we can move content between line#218 and line#225 to for-block.

Reply:
The original design of this block is to test re-downloading truncated/deleted files.
At the start of each round, we downloaded all files up to targetTs.
But cleaning up and re-download is cleaner.
Fixed.